compiler/rustc_next_trait_solver/src/solve/trait_goals.rs

```diff
@@ -1288,13 +1288,20 @@ where
             }
             _ => false,
         });
-        if has_non_global_where_bounds {
+
+        // As mentioned above, if there are non-global where bounds, then prefer all where bounds
+        // (including global ones) over anything else. If there are *only* where bounds, then
+        // make sure to return that this is proven-via the param-env so that rigid projections
+        // operate correctly.
+        if has_non_global_where_bounds
+            || (candidates.len() > 0
+                && candidates.iter().all(|c| matches!(c.source, CandidateSource::ParamEnv(_))))
```

Comment:
While current, the controlflow feels a bit confusing to me. I think I'd prefer to instead add the following the "merge all candidates" success branch
```rust
let proven_via = if candidates.iter().all(|c| matches!(c.source, CandidateSource::ParamEnv(_)) { TraitGoalProvenVia::ParamEnv } else { TraitGoalProvenVia::Misc };
```
with that, r=me